[Videoversions] Extend video version functionality to allow the selection of multiple versions (playlists) in the same Blu-ray ISO/folder - and add support for import/export of (multi) movie versions.

78andyp opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Now that bluray:// file paths are implemented to point to individual playlists this PR extends the video version functionality to allow the selection of multiple playlsits on the same disc and assign each a version.

This also extends import/export functionality to support multi-movie files (ie. blurays in the case) - but could be easily extended to include mkv files with more than one movie in the future. It also fixes some existing bugs regarding movie versions import/export.

For multiple movies in a file (ie bluray), multiple `<movie>` entries are created in the .nfo - all encapsulated in a root <movies> to ensure xml compliance - one for each movie version. This because there is only one base file (.ISO or index.bdmv) and the playlist information is stored in the nfo itself - and the interaction between VideoInfoScanner and the NFO classes makes it difficult to scan all nfo files and communicate back (the nfo file is 'found' in the NFO classes not in VideoInfoScanner).

How to use:
When the version dialog detects that you are managing versions on a bluray it asks first if you want to assign a playlist to the default version (if not played already) then if you want to add an additional version/playlsit on the disc. If yes, then the simple menu dialog is used to select a playlist. If no, then the usual library/browse options are presented.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Currently you cannot select multiple video versions within the same file, so for blurays where you have multiple versions (eg. Terminator 2 - Theatrical version, Special Edition and Extended Edition) on the same disc you cannot use this feature.

**Failure to import movie versions**

Firstly create two versions of a movie with media files.

![image](https://github.com/user-attachments/assets/36b05b72-261d-4477-a310-da47890a6c77)

Export -> Separate -> Yes to art -> Overwrite
Delete database
Set content Movies -> Local Scraper -> Individual folders

![image](https://github.com/user-attachments/assets/07adfa50-f51f-4678-801e-323f8b97bf05)

Note versions are not created.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Locally

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

![image](https://github.com/user-attachments/assets/a39d23f8-a70b-453d-b092-68b64f0d7ca7)

![image](https://github.com/user-attachments/assets/eb2e9a1a-463b-4be1-b6cd-0c550865d1af)

![image](https://github.com/user-attachments/assets/b2034dad-6daa-4fbd-90b0-49b00500dd6b)

More complex test:

```
+---Aliens (1986)
    ¦   Aliens - The Cocoon Cut-poster.png
    ¦   Aliens - The Cocoon Cut.mp4
    ¦   Aliens - The Cocoon Cut.nfo
    ¦   
    +---Disc 1
    ¦       ALIENS.dvd
    ¦       ALIENS.iso
    ¦       ALIENS.md5
    ¦       
    +---Disc 2
        ¦   disc.inf
        ¦   
        +---BDMV
        ¦   ¦   index.bdmv
        ¦   ¦   MovieObject.bdmv
 ```

The Bluray discs are taken from the UHD release. Disc 1 is 4K - containing both the theatrical and special editions. Disc 2 is bluray - again containing both editions. The Cocoon cut is a fan edit.

Adding the source using the TMDB scraper:
![image](https://github.com/user-attachments/assets/81fd1065-41f4-4f24-b0d4-8c21cee8b5cd)

On Aliens (Disc 1) -> Context menu -> Manage.. -> Manage versions
Add version
Yes to assign playlist to current movie version -> 800
Yes to add another version from this disc -> 801 -> Special Edition
Add version
No to add another version from this disc
Select Aliens (Disc 2) from the dialog
Select title 800 -> Standard edition

![image](https://github.com/user-attachments/assets/7eb5f11d-b34e-40ab-b3d0-1f8559919a1c)

Now left click/select the bottom Standard edition (on Disc 2) entry

NB - this is how add version knows which disc to look for a version on.

![image](https://github.com/user-attachments/assets/9b2517ef-a60e-4528-b040-498357035cf5)

Add version
Yes to add another version from this disc -> 801 -> Special Edition

![image](https://github.com/user-attachments/assets/1fa72e93-1220-48cb-8dc3-48aed0077ea3)

Add version
No to add another version from this disc
Browse Library
Select Aliens – The Cocoon Cut -> Fan edit

Now the versions manager looks like:

![image](https://github.com/user-attachments/assets/2b37a050-5398-4a66-96ba-94035e3b5154)

Now the library looks like:

![image](https://github.com/user-attachments/assets/81c562aa-276a-4f01-bcd8-625167c3d003)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

78andyp committed:
Get stream details at playlist selection.

Get streamdetails from BLURAY_TITLE_INFO during playlist selection. These may not be completely accurate so ensure these are then updated when the playlist is actually played.

78andyp commented:
@ksooo - grateful for your thoughts. Thanks.

78andyp committed:
Introduce noremoteartwithlocalscraper Advanced Setting.

Currently, even if the local information only scraper is selected, if a local NFO contains links to online art - this online art is used.

If set to true (default false) - this setting mean no online art will be used at all, even if present in a local NFO.

Most of the code changes are simply to pass scraper information through the functions.

CrystalP commented:
It's not an easy problem to tackle (this is only part of the whole Bluray disc support) and tbh I don't know how widespread the need is, since the most frequent reason for users to not remux the discs as mkv is the ability to use the menus and other niceties (or a lack of expertise / time / motivation I guess).

Do you have ideas for UI design beyond this PR? I'm not sure the approach would scale when extended for the rest of the problem of Bluray discs support. Off the top of my head, discs can also contain different movies, extras, music videos.

It may be possible to solve the problem addressed by the PR with multiple .strm files, each pointing to a different playlist?

From a technical pov, the PR allows multiple versions (same or different movie parents) to point to the same playlist (and file record), which is very likely to break things as the file id acts as a primary key.
Not sure why the precautions around StreamDetails?

78andyp commented:
> It's not an easy problem to tackle (this is only part of the whole Bluray disc support) and tbh I don't know how widespread the need is, since the most frequent reason for users to not remux the discs as mkv is the ability to use the menus and other niceties (or a lack of expertise / time / motivation I guess).
> 
> Do you have ideas for UI design beyond this PR? I'm not sure the approach would scale when extended for the rest of the problem of Bluray discs support. Off the top of my head, discs can also contain different movies, extras, music videos.
> 
> It may be possible to solve the problem addressed by the PR with multiple .strm files, each pointing to a different playlist?
> 
> From a technical pov, the PR allows multiple versions (same or different movie parents) to point to the same playlist (and file record), which is very likely to break things as the file id acts as a primary key. Not sure why the precautions around StreamDetails?

Hi

Good points. I'l try and tackle them in turn.

Agree this is just part of bringing disc (bluray and then dvd) support up to be on par with file support. I've most of the pieces in various stages of development. I outlined somewhere else, but am happy to summarise again if that's helpful.

It's difficult to say how much demand there is. On the one hand - this is currently still pre-alpha, so most people here are probably more than capable of muxing their own mkvs or just putting up with it. There are certainly still people out there who like whole discs, and I know some versions like LibreElec only get updated once gone gold. There is a space issue as well (one UHD v. 3 demuxed MKVs of nearly the same same for my T2 disc with 3 editions for example).

How much work should we put in to niche features I don't konw - but there clearly was some intent (eg. the existence of bluray:// vfs paths etc..) to introduce this sort of feature at some point. Are discs on the way out now we have streaming - a whole other debate!!

I disagree re: scale. This approach would scale. Each extra etc.. would be a seperate playlist - and each playlist is only one item (movie version/extra etc..) I agree it needs a UI to allow playlist <-> extra matching but I think that could be done using that which is already there. It is on my list of things to do, I'd just like to get this and epiosdes (uploading shortly) signed off first. I can't see a need for strm files - although more than happy to stand corrected.

You are right about the multiple playlists - I need to extend the playlist selection dialog to allow exclusions of playlists already selected.

Cheers
A

78andyp commented:
> It's not an easy problem to tackle (this is only part of the whole Bluray disc support) and tbh I don't know how widespread the need is, since the most frequent reason for users to not remux the discs as mkv is the ability to use the menus and other niceties (or a lack of expertise / time / motivation I guess).
> 
> Do you have ideas for UI design beyond this PR? I'm not sure the approach would scale when extended for the rest of the problem of Bluray discs support. Off the top of my head, discs can also contain different movies, extras, music videos.
> 
> It may be possible to solve the problem addressed by the PR with multiple .strm files, each pointing to a different playlist?
> 
> From a technical pov, the PR allows multiple versions (same or different movie parents) to point to the same playlist (and file record), which is very likely to break things as the file id acts as a primary key. Not sure why the precautions around StreamDetails?

Prevented selection of the same playlist twice by adding exclusion list.

Streamdetails - in the first place I avoided updating them as there was know way of knowing which playlist was selected if the disc was played prior to the bluray:// PRs. However with the last commit here, where streamdetails are fetched from the dics itself - they will get updated anyway so I've simplified it.

78andyp committed:
Fix only show Choose Playlist on context menu when not more appropriate to select version playlist through version manager dialog.

ksooo commented on xbmc/filesystem/BlurayFile.h:
```diff
@@ -10,18 +10,18 @@
 #include "filesystem/OverrideFile.h"
 
-namespace XFILE
+using namespace XFILE;
```
No "using namespace" in header files, please!  This can lead to weird compile problems.  

More important: You here also define the namespace for the class defined here! With your change, `CBlurayFile` no longer is in namespace `XFILE`, but it should.

Please refrain  from making changes which are definitely not related to the subject of the commit message. Make a separate commit so such things can easily be spotted and reviewed.

ksooo reviewed (commented):
Only had a look at the first commit...

78andyp committed:
Add scrolling to decoded path in version manager dialog.

78andyp committed:
Use parent art as fallback in video version manager dialog.

yol commented:
jenkins build this please

78andyp committed:
Export movie versions (including blurays).

This is achieved by creating multiple <movie> entries in the .nfo - all encapsulated in a root <movies> to ensure xml compliance - one for each movie version. This because there is only one base file (.ISO or index.bdmv).

If there are multiple movie versions each with a different filename then multiple .nfo files are created (current behaviour).

78andyp committed:
Import movie versions (including blurays).

This also addresses the existing bug where video versions are not correctly imported from separate .nfos.

CVideoDatabase::GetMovieId() is modified to allow the option of matching on the basis of directory and not full file (to show that there is already a version in the directory in the library) and GetMovieByTitle() is introduced.

Routines are created in CVideoDatabase to add movie versions without creating a new movie.

This also fixes a bug whereby an import did not properly set the video version type (as AddVideo() assumed it was Standard Edition).

78andyp committed:
Add ability to choose playlist for video versions.

Uses the simple menu and then updates the database to reflect the playlist choice, creating a new videoversion table entry when needed.

garbear commented:
I gave the diff a look-over and didn't notice any problems.

78andyp commented:
> I gave the diff a look-over and didn't notice any problems.

Thanks :)

78andyp commented:
Final call for any feedback on this. Will merge tomorrow.

heitbaum commented:
> Hi @mackal. Could you confirm #27036 fixes it for you. Thanks.

Build now successful.

garbear commented:
Thanks for the fast fix!
